gobblin-metrics/src/test/java/gobblin/metrics/KafkaReporterTest.java

```diff
@@ -0,0 +1,228 @@
+    expectedSet.add("com.linkedin.example.histogram.999percentile");
+    expectedSet.add("com.linkedin.example.histogram.count");
+
+    expectMetrics(iterator, expectedSet, true);
+
+    kafkaReporter.close();
+
+  }
+
+  @Test
+  public void kafkaReporterTagsEnvHostTest() throws IOException {
+    MetricRegistry registry = new MetricRegistry();
+    Counter counter = registry.counter("com.linkedin.example.counter");
+
+    String host = "host.linkedin.com";
+    String env = "testing";
+    String tag1 = "tag1";
+    String tag2 = "tag2";
+
+    KafkaReporter kafkaReporter = getBuilder(registry).
+        withHost(host).
+        withEnv(env).
+        withTags(tag1, tag2).
+        build("localhost:" + kafkaPort, _topic);
+
+    counter.inc();
+
+    kafkaReporter.report();
+
+    try {
+      Thread.sleep(100);
+    } catch(InterruptedException ex) {
+      Thread.currentThread().interrupt();
+    }
+
+    KafkaReporter.Metric metric = nextMetric(iterator);
+
+    Assert.assertEquals(1, Integer.parseInt(metric.value.toString()));
+    Assert.assertEquals(host, metric.host);
+    Assert.assertEquals(env, metric.env);
+    Assert.assertEquals(2, metric.tags.size());
+    Assert.assertTrue(metric.tags.contains(tag1));
+    Assert.assertTrue(metric.tags.contains(tag2));
+
+  }
+
+  /**
+   * Expect a list of metrics with specific values.
+   * Fail if not all metrics are received, or some metric has the wrong value.
+   * @param it ConsumerIterator for Kafka topic
+   * @param expected map of expected metric names and their values
+   * @throws IOException
+   */
+  private void expectMetricsWithValues(ConsumerIterator<byte[], byte[]> it, Map<String, Double> expected)
+      throws IOException {
+    System.out.println("====Checking for expected messages and values. Will list all messages.====");
+    try {
+      while(it.hasNext()) {
+        KafkaReporter.Metric metric = nextMetric(it);
+        if (expected.containsKey(metric.name)) {
+          Assert.assertEquals(expected.get(metric.name), Double.parseDouble(metric.value.toString()));
+          expected.remove(metric.name);
+        }
+      }
+    } catch (ConsumerTimeoutException e) {
+      Assert.assertTrue(expected.isEmpty());
+    }
+  }
+
+  /**
+   * Expect a set of metric names. Will fail if not all of these metrics are received.
+   * @param it ConsumerIterator for Kafka topic
+   * @param expected set of expected metric names
+   * @param strict if set to true, will fail if receiving any metric that is not expected
+   * @throws IOException
+   */
+  private void expectMetrics(ConsumerIterator<byte[], byte[]> it, Set<String> expected, boolean strict)
+      throws IOException {
+    System.out.println("====Checking for expected messages. Will list all messages.====");
+    try {
+      while(it.hasNext()) {
+        KafkaReporter.Metric metric = nextMetric(it);
+        //System.out.println(String.format("expectedSet.add(\"%s\")", metric.name));
```

Comment:
I left this here because the line is useful when adding more unit tests (you can uncomment that, and it will print the code to add all expected metrics, instead of having to write them by hand). I can either add comment explaining why it's there, or completely remove it. I'm fine with either.